### PR TITLE
Speed up Building Indices Lookup in Metadata

### PR DESCRIPTION
**Pull request by original-brownbear**
We can speed this up by almost a third in case of a large index count
where most indices are part of a datastream by avoiding the double lookup
for the ds abstraction in the way done here.
Also, simplified the loop iteration a little to use the slightly faster cursor
and removed some needless conditional in the loop.

relates #77466

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 1]
```diff
@@ -1817,35 +1817,27 @@ public Metadata build(boolean builtIndicesLookupEagerly) {
...
-                } else {
-                    index = new ConcreteIndex(indexMetadata);
-                }
-
-                IndexAbstraction existing = indicesLookup.put(indexMetadata.getIndex().getName(), index);
+            for (var entry : indices) {
+                final String name = entry.key;
+                final IndexMetadata indexMetadata = entry.value;
+                final IndexAbstraction.DataStream parent = indexToDataStreamLookup.get(name);
+                assert parent == null || parent.getIndices().stream().anyMatch(index -> name.equals(index.getName()))
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1838]: assertion change is unrelated, just felt the need to clean this up and make it a little clearer what we're asserting :)

Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1838]: totally a nit and feel free to ignore, but since we are using streams, I would prefer to keep `.map(Index::getName)`, i.e.:

```suggestion
                assert parent == null || parent.getIndices().stream().map(Index::getName).anyMatch(name::equals)
```

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 2]
```diff
@@ -1817,35 +1817,27 @@ public Metadata build(boolean builtIndicesLookupEagerly) {
...
+                    final IndexAbstraction.DataStream dsAbstraction = new IndexAbstraction.DataStream(dataStream, aliases);
                     IndexAbstraction existing = indicesLookup.put(
                         dataStream.getName(),
                         new IndexAbstraction.DataStream(dataStream, aliases)
                     );
                     assert existing == null : "duplicate data stream for " + dataStream.getName();
 
                     for (Index i : dataStream.getIndices()) {
-                        indexToDataStreamLookup.put(i.getName(), dataStream);
+                        indexToDataStreamLookup.put(i.getName(), dsAbstraction);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1828]: I wonder if we need to build `indextToDataStreamLookup` at all? Could we not simply populate `indicesLookup` directly here instead and ignore those below (asserting that they are data stream owned instead of the duplicate check)?

**Review by henningandersen — APPROVED** [review]
LGTM.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1828]: Right we don't ... now comes the interesting question of whether this would actually be faster. 

Looking at the profiling for the full method (after my change here):

image

for a completely degenerate case where we have about as many indices as datastreams (one index per DS) ...
the hash map building barely shows up relative to the tree-map. 

So the question becomes, is it cheaper to do 2 puts (or I guess one put and one computeIfAbsent) to the tree-map (the second put in the indices loop would be to a larger tree map than the one in the DS loop) or do one put into an on average smaller tree map and instead build the technically redundant hash-map.

Profiling suggests that this is faster I'd say. And in the real world, where you'd have multiple indices per DS the difference would be even more visible.
The problem we're running into is that the tree-map is super lob sided because we have the shared `.ds` prefix for all indices pretty much and then some more sharing for the specific DS, leading to loads of comparisons.
We should've used a different naming schema for the internal DS indices I guess ... (in fact I think we still could).

**Comment by original-brownbear** [comment]
Thanks Henning! I went with my solution for now because of the way our tree-map is painful. But I think we might want to improve that in a follow-up.
